library/alloc/src/vec/mod.rs

```diff
@@ -1265,6 +1288,18 @@ impl<T, A: Allocator> Vec<T, A> {
     /// }
     /// assert_eq!(&*x, &[0, 1, 2, 3]);
     /// ```
+    ///
+    /// The validity guarantee works out this way:
+    ///
+    /// ```rust
+    /// let mut v = vec![0];
+    /// let ptr1 = v.as_mut_ptr();
+    /// ptr1.write(1);
+    /// let ptr2 = v.as_mut_ptr();
+    /// ptr2.write(2);
+    /// // Notably, the write to `ptr2` did *not* invalidate `ptr1`:
+    /// ptr1.write(3);
+    /// ```
```

Comment:
There's another example, inspired by you as_ptr example:
```rust
let mut v = vec![0];
let ptr1 = v.as_mut_ptr();
v[0] = 1;
ptr1.read(); // UB or not?
```
For `addr_of_mut`, we have *not* committed to actually making this legal, so we might not want to commit to it here either.

OTOH if we do another `as_mut_ptr` after the write then reading from that is obviously fine. So... by analogy with `addr_of_mut`, maybe saying that the pointers are identical for the aliasing model isn't right. Maybe it should be more like, "two raw pointers obtained via two separate calls to this method can be used in an arbitrary interleaved fashion, i.e. they do not have aliasing restrictions with respect to each other (but they still *can* have aliasing restrictions with respect to accesses done via other means, such as indexing into the vector)".

Reply:
Yeah, that's kinda what I was thinking.

Reply:
Actually, going back to https://github.com/thomcc/rust-typed-arena/issues/62 , there has to be some amount of other operations we allow on the vec for that kind of code to be safe. The rule seems to be "do not materialize `&`-refs to the entire slice or to this element in the slice through any method that is not `as_mut_ptr()` APIs"?

Reply:
Yes, basically -- but then that begs the question which operations materialize which references. For instance, indexing materializes a reference *to that particular element* (that's why my example above is not allowed), but not to other elements.

Reply:
We could potentially list a _couple_ things that are allowed, and use terminology about materializing references that let you ask questions of the libs team in the future if needed?

Like, we say "this method does not itself materialize a reference and will return a pointer that is valid as long as other code does not materialize a (mutable, immutable) reference to the entire slice or to this particular element except through this pointer", and then if someone is curious if `len()` counts, they can ask if it materializes a reference.

Reply:
That works for me.

Reply:
(I would want to be careful to not like ... explain half of stacked/tree/graph/heap borrows in these docs but i think the framing i have put forth so far strikes an okay balance)